core/src/replay_stage.rs

```diff
@@ -850,36 +851,63 @@ impl ReplayStage {
         }
     }
 
+    fn maybe_retransmit_unpropagated_slots(
+        metric_name: &'static str,
+        retransmit_slots_sender: &RetransmitSlotsSender,
+        progress: &mut ProgressMap,
+        latest_leader_slot: Slot,
+    ) {
+        let first_leader_group_slot = first_of_consecutive_leader_slots(latest_leader_slot);
+
+        for slot in first_leader_group_slot..=latest_leader_slot {
+            if !progress.is_propagated(slot).unwrap_or(true) {
+                if let Some(retransmit_info) = progress.get_retransmit_info(slot) {
```

Comment:
For future reference, maybe a comment on a situation in which this check could fail, illustrated by this proposed test case here: https://github.com/solana-labs/solana/pull/22157/files#r777720976

Something along the lines of "If a leader skipped one of their scheduled slots `S`, then the progress map will not contain the slot `S`, and the `progress.is_propagated(slot).unwrap_or(true)` check above will still return true. But the existence check will then fail here.

Reply:
Might also make sense to reverse the order of these checks like:

```
if let Some(retransmit_info) = progress.get_retransmit_info(slot) {
   if !progress.is_propagated(slot).expect("Above check ensures progress map must contain this entry");
}

```

Reply:
good call, switched the ordering